### PR TITLE
feat(plugin): WebhookTags -> MoreUserTags

### PR DESCRIPTION
**Pull request by ActuallyTheSun**
alongside tags for webhooks, this plugin adds tags for moderative roles
chat-message
also adding them to the sidebar and profiles
member-list
profile
and it combines bot and op tags so they both show
bot-admin
op-owner

### CODE REVIEW

**Comment by C10udburst** [comment]
https://github.com/ActuallyTheSun/Vencord/pull/1 <:verycool:888951267873923093>

**Comment by ActuallyTheSun** [comment]
real

File: src/utils/patches.ts [hunk 1]
```diff
@@ -29,7 +29,7 @@ export function canonicalizeMatch(match: RegExp | string) {
 
 export function canonicalizeReplace(replace: string | ReplaceFn, pluginName: string) {
     if (typeof replace === "function") return replace;
-    return replace.replaceAll("$self", `Vencord.Plugins.plugins.${pluginName}`);
+    return replace.replaceAll("$self", pluginName.includes(" ") ? `Vencord.Plugins.plugins["${pluginName}"]` : `Vencord.Plugins.plugins.${pluginName}`);
```
Comment (Vendicated) [inline src/utils/patches.ts:32]:
```suggestion
    return replace.replaceAll("$self", `Vencord.Plugins.plugins["${pluginName}"]`);
```

**Comment by C10udburst** [comment]
@Vendicated will merge

**Comment by KraXen72** [comment]
+1 plz merg

**Comment by castdrian** [comment]
the option to use the round OP tags for all would be great + maybe different colors per permission

File: src/plugins/moreTags.ts [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+import definePlugin, { OptionType } from "@utils/types";
+import { waitFor } from "@webpack";
+import { ChannelStore, GuildStore } from "@webpack/common";
+import { Channel, Message, User } from "discord-types/general";
+
+import { Settings } from "../Vencord";
+
+let Permissions: Record<string, bigint>, computePermissions: ({ ...args }) => bigint, Tags: Record<string, number>;
+waitFor(["SEND_MESSAGES", "VIEW_CREATOR_MONETIZATION_ANALYTICS"], m => Permissions = m);
+waitFor(["computePermissions", "canEveryoneRole"], m => ({ computePermissions } = m));
```
Comment (Vendicated) [inline src/plugins/moreTags.ts:30]: Isn't this already in webpack Commons as PermissionStore?

File: src/plugins/moreTags.ts [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+const passChannelIdDownProps = [
+    {
+        find: ".hasAvatarForGuild(null==",
+        replacement: {
+            match: /\(\).usernameSection,user/,
+            replace: "().usernameSection,moreTags_channelId:arguments[0].channelId,user"
+        }
+    },
+    {
+        find: "().copiableNameTag",
```
Comment (Vendicated) [inline src/plugins/moreTags.ts:92]: please update these patches to latest canary change! these modules no longer use functions, so the () must be removed

File: src/plugins/moreTags.ts [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        replacement: {
+            match: /discriminatorClass:(.{1,100}),botClass:/,
+            replace: "discriminatorClass:$1,moreTags_channelId:arguments[0].moreTags_channelId,botClass:"
+        }
+    }
+];
+
+migratePluginSettings("More Tags", "Webhook Tags");
+export default definePlugin({
+    name: "More Tags",
```
Comment (Vendicated) [inline src/plugins/moreTags.ts:102]:
```suggestion
    name: "MoreTags",
```

File: src/plugins/moreTags.ts [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    description: "Adds tags for webhooks and moderative roles (owner, admin, etc.)",
+    authors: [Devs.Cyn, Devs.TheSun],
+    options: {
+        dontShowBotTag: {
+            description: "Don't show \"BOT\" text for bots with other tags (verified bots will still have checkmark)",
```
Comment (Vendicated) [inline src/plugins/moreTags.ts:107]:
```suggestion
            description: 'Don't show "BOT" text for bots with other tags (verified bots will still have checkmark)',
```

File: src/plugins/moreTags.ts [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+                {
+                    match: /case (\i)\.BOT:default:(\i)=(.{1,20})\.BOT/,
+                    replace: (orig, types, text, strings) =>
+                        `${tags.map(t =>
+                            // case r.SERVER: T = c.Z.Messages.BOT_TAG_SERVER; break;
+                            `case ${types}.${t.name}:${text}="${t.displayName}";break;\
+${t.botAndOpCases ? `case ${types}.${t.name}_OP:${text}=${strings}.BOT_TAG_FORUM_ORIGINAL_POSTER+" • ${t.displayName}";break;\
+case ${types}.${t.name}_BOT:${text}=${strings}.BOT_TAG_BOT+" • ${t.displayName}";break;` : ""}`
+                        ).join("")}${orig}`
+                },
```
Comment (Vendicated) [inline src/plugins/moreTags.ts:160]: these patches are terribly unreadable, please make them more readable somehow, or why not extract this logic into a method and call that instead?

File: src/plugins/moreTags.ts [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+                match: /,botType:(\i\((\i)\)),/,
+                replace: ",botType:$self.getTag({user:$2,channelId:arguments[0].moreTags_channelId,origType:$1,location:'not-chat'}),"
+            }
+        },
+        ...passChannelIdDownProps
+    ],
+
+    getPermissions(user: User, channel: Channel): string[] {
+        if (!channel) return [];
+        const guild = GuildStore.getGuild(channel.guild_id);
```
Comment (Vendicated) [inline src/plugins/moreTags.ts:198]:
```suggestion
        const guild = GuildStore.getGuild(channel?.guild_id);
```
if you do this, you can delete the above channel check

File: src/plugins/moreTags.ts [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+        if (!guild) return [];
+        // PermissionStore.computePermissions is not the same function and doesn't work here
+        const permissions = computePermissions({ user, context: guild, overwrites: channel.permissionOverwrites });
+        return Object.entries({ ...Permissions }).map(([perm, permInt]) =>
+            permissions & permInt ? perm : ""
+        ).filter(i => i);
```
Comment (Vendicated) [inline src/plugins/moreTags.ts:204]:
```suggestion
        ).filter(Boolean);
```

File: src/plugins/moreTags.ts [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+        ...passChannelIdDownProps
+    ],
+
+    getPermissions(user: User, channel: Channel): string[] {
+        if (!channel) return [];
+        const guild = GuildStore.getGuild(channel.guild_id);
+        if (!guild) return [];
+        // PermissionStore.computePermissions is not the same function and doesn't work here
+        const permissions = computePermissions({ user, context: guild, overwrites: channel.permissionOverwrites });
+        return Object.entries({ ...Permissions }).map(([perm, permInt]) =>
```
Comment (Vendicated) [inline src/plugins/moreTags.ts:202]: whats the point of this spread?

File: src/plugins/moreTags.ts [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+            permissions & permInt ? perm : ""
+        ).filter(i => i);
+    },
+
+    // custom tags have id's of minimum 100, being xx0 for normal, xx1 for op and xx2 for bot
+    // for example: 100 101 102, 110 111 112, 120 121 122, etc.
+    isOPTag: (tag: number) => tag === Tags.ORIGINAL_POSTER || (tag >= 100 && tag % 10 === 1),
```
Comment (Vendicated) [inline src/plugins/moreTags.ts:209]: wtf

**Review by Vendicated — CHANGES_REQUESTED** [review]

File: src/plugins/moreTags.ts (hunk 9 above)
Comment (KraXen72) [reply src/plugins/moreTags.ts:202]: maybe a shallow copy?

File: src/plugins/moreTags.ts (hunk 2 above)
Comment (Vendicated) [reply src/plugins/moreTags.ts:30]: ?

Comment (ActuallyTheSun) [reply src/plugins/moreTags.ts]: oh i forgot to give a reason

`computePermissions` from PermissionStore isnt the same function and that one cant be used here (the one from PermissionStore can only compute for the current user). i've added a comment to clear things up

**Comment by ActuallyTheSun** [comment]
maybe rename to MoreUserTags or StaffTags to avoid confusion with MessageTags?

File: src/plugins/moreTags.ts [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+                {
+                    match: /(\i)\[.\.BOT=0\]="BOT";/,
+                    replace: (orig, types) =>
+                        `${tags.map(t =>
+                            `${addTagVar(t.name, types)};${addTagVar(`${t.name}_OP`, types)};${addTagVar(`${t.name}_BOT`, types)};`
+                        ).join("")}${orig}`
+                },
+                {
+                    match: /case (\i)\.BOT:default:(\i)=(.{1,20})\.BOT/,
+                    replace: (orig, types, text, strings) =>
```
Comment (Vendicated) [inline src/plugins/moreTags.ts:142]: i don't like these two replacements. You're dynamically building a lot of javascript. Why not use a proper function for this? That's so much more robust

File: src/plugins/moreTags.ts [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+            if (
+                tag.permissions?.some(perm => perms.includes(perm)) ||
+                (tag.condition?.(message, user, channel))
+            ) {
+                if (channel.isForumPost() && channel.ownerId === user.id)
+                    type = Tags[`${tag.name}_OP`];
+                else if (user.bot && !isWebhook(message, user) && !settings.dontShowBotTag)
+                    type = Tags[`${tag.name}_BOT`];
+                else
+                    type = Tags[tag.name];
```
Comment (Vendicated) [inline src/plugins/moreTags.ts:253]: if you add `break` to these if cases, you can remove the .reverse() above and do less work, no?

**Review by Vendicated — CHANGES_REQUESTED** [review]

**Pull request merged**